File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    set<int> rows[9];
    set<int> columns[9];
    set<int> boxes[9];

    bool isValidSudoku(vector<vector<char>>& board) {
        
        for (int i=0; i<9; i++) {
            for (int j=0; j<9; j++) {
                int num = board[i][j];
                
                if (num == '.') {
                    continue;
                }

                if (rows[i].count(num) || columns[j].count(num) || boxes[(i/3)*3 + j/3].count(num)) {
                    return false;
                }

                rows[i].insert(num);
                columns[j].insert(num);
                boxes[(i/3)*3 + j/3].insert(num);
            }
        }

        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        n = len(board)
        hash_map = defaultdict(bool)

        for i in range(n):
            for j in range(n):
                if board[i][j] == ".":
                    continue
                if hash_map[board[i][j]]:
                    return False
                hash_map[board[i][j]] = True
            
            hash_map.clear()

        for i in range(n):
            for j in range(n):
                if board[j][i] == ".":
                    continue
                if hash_map[board[j][i]]:
                    return False
                hash_map[board[j][i]] = True

            hash_map.clear()
        
        for i in range(0, n, 3):
            for j in range(0, n, 3):
                for l in range(i, i+3):
                    for k in range(j, j+3):
                        if board[l][k] == ".":
                            continue
                        if hash_map[board[l][k]]:
                            return False
                        hash_map[board[l][k]] = True
                    
                hash_map.clear()
        
        return True

